### DISCUSSION TITLE
postgres connection failing when password has a $ in it

### DISCUSSION THREAD

**Discussion by benpreisler** [sqlalchemy/sqlalchemy#6615] [Usage Questions] [answered]
This is baffling me so I thought I'd throw it out and see if anyone else has run into this or can recreate it.

I've had a db password with a $ in it for months and yesterday I updated from sqlalchemy 1.4.14 to 1.4.17 and started getting exceptions on the connection string.  Given a password of my$pwd I get the following out on the server:

```
  File "/opt/python/lib/python3.8/site-packages/psycopg2/__init__.py", line 127, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not translate host name "$pwd@dbname" to address: Name or service not known
```

which looks like a parsing problem.

And trying it locally against a postgres instance in a docker container I get a different but similar error:

```
        dsn = _ext.make_dsn(dsn, **kwargs)
>       conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
E       sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  password authentication failed for user "postgres"
```

locally I'm using:
sqlalchemy==1.4.14
psycopg2-binary==2.8.6
postgres=12.7 (but 12.5 and 12.6 have the same issue)

Tracing it down sqlalchemy seems to be parsing the password just fine and it only fails once it hits:
`connection = pool._invoke_creator(self)`
here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/pool/base.py#L605

So it seems like its failing in the driver...  (but I haven't changed that recently)

locally my connection string looks like this:
postgresql://postgres:my$pwd@127.0.0.1:5001/my_db

Is anyone else able to reproduce this?

**Comment by zzzeek** [comment, marked as answer]
make sure you use URL escaping for passwords:

https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls

some issues in URL parsing were fixed a few 1.4 versions ago which unfortunately caused some unavoidable regressions in some passwords.

**Reply by benpreisler** [reply]
ok thanks - good to know

I was confused because 1.4.14 worked - moving to 1.4.17 broke - but moving back to 1.4.14 was still broken.  Oh well it's fixed now - thanks!

**Reply by matletix** [reply]
You can use urllib for encoding the password : 
```python
from urllib.parse import quote_plus

encoded_pwd = quote_plus(raw_password)
```

**Reply by CaselIT** [reply]
@zzzeek maybe we could add that snipped in the URL docs?

**Reply by zzzeek** [reply]
we do have that snippet in https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls  ?

**Reply by CaselIT** [reply]
we do indeed, but maybe we could link it also from the url / make_url docs https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.engine.URL

**Reply by zzzeek** [reply]
OK

**Reply by CaselIT** [reply]
tried adding something here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3754 but it may be a bit too much with all these dragons

**Reply by zzzeek** [reply]
I would use ..seealso without any specific warning for most of those boxes.  I would put "..note" only under the `:param password:` description.

Overall, this is not a long term issue.  the reason we keep hearing about it is because we made a backwards incompatible change in the URL scheme so we are getting people who are upgrading.     We are already getting fewer of these and eventually it will be a non-issue.

**Reply by CaselIT** [reply]
updated the gerrit to only have see also
